pydantic/_internal/_generate_schema.py

```diff
@@ -858,27 +862,33 @@ def _resolve_forward_ref(self, obj: Any) -> Any:
         return obj
 
     @overload
-    def _get_args_resolving_forward_refs(self, obj: Any, required: Literal[True]) -> tuple[Any, ...]: ...
+    def _get_args_resolving_forward_refs(
+        self, obj: Any, *, required: Literal[True], eval_str: bool = True
+    ) -> tuple[Any, ...]: ...
 
     @overload
-    def _get_args_resolving_forward_refs(self, obj: Any) -> tuple[Any, ...] | None: ...
+    def _get_args_resolving_forward_refs(self, obj: Any, *, eval_str: bool = True) -> tuple[Any, ...] | None: ...
 
-    def _get_args_resolving_forward_refs(self, obj: Any, required: bool = False) -> tuple[Any, ...] | None:
+    def _get_args_resolving_forward_refs(
+        self, obj: Any, *, required: bool = False, eval_str: bool = True
+    ) -> tuple[Any, ...] | None:
         args = get_args(obj)
         if args:
+            if eval_str:
+                args = [_typing_extra._make_forward_ref(a) if isinstance(a, str) else a for a in args]
```

Comment:
@sydney-runkle @Viicos How about this pattern?

Reply:
I'm not a fan of the added logic here. I think it's safe enough to do:

```python
args = tuple([self._resolve_forward_ref(a) if isinstance(a, (ForwardRef, str)) else a for a in args])
```

This will not behave as expected for `Literal` though, but probably we can add a `assert not is_literal_type(obj), 'obj must not be a Literal instance'` just so that we avoid mistakenly using this method.

Reply:
Unfortunately `self._resolve_forward_ref` does not resolve `str` objects because under the hood it calls `_typing_extra.eval_type_backport` instead of `_typing_extra.eval_type`. I guess we have to use `_typing_extra._make_forward_ref` anyways.